File: code/python/algorithm_thinking_article-16.py
"""Fancy traversal techniques for two-dimensional arrays"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().reverseWords(*args, **kwargs)


class _Solution:
    def reverseWords(self, s: str) -> str:
        sb = []
        for c in s:
            if c != ' ':
                sb.append(c)
            elif sb and sb[-1] != ' ':
                sb.append(' ')
        if sb and sb[-1] == ' ':
            sb.pop()
        chars = sb
        def reverse(l, r):
            while l < r:
                chars[l], chars[r] = chars[r], chars[l]
                l += 1
                r -= 1
        n = len(chars)
        reverse(0, n-1)
        start = end = 0
        while end < n:
            if chars[end] == ' ':
                reverse(start, end-1)
                start = end + 1
            end += 1
        reverse(start, n-1)
        return "".join(chars)

File: code/javascript/algorithm_thinking_article-16.js
/**
 * Fancy traversal techniques for two-dimensional arrays
 */
export function solve(...args) {
  // Entry point for the course exercise: pass inputs directly.
  return reverse(...args);
}

var reverseWords = function(s) {
    let sb = "";
    for (let i = 0; i < s.length; i++) {
        let c = s.charAt(i);
        if (c !== ' ') {
            sb += c;
        } else if (sb !== "" && sb[sb.length - 1] !== ' ') {
            sb += ' ';
        }
    }
    if (sb[sb.length - 1] === ' ') {
        sb = sb.substring(0, sb.length - 1);
    }
    let chars = sb.split('');
    let n = chars.length;
    reverse(chars, 0, n - 1);
    for (let i = 0; i < n;) {
        for (let j = i; j < n; j++) {
            if (j + 1 === n || chars[j + 1] === ' ') {
                reverse(chars, i, j);
                i = j + 2;
                break;
            }
        }
    }
    return chars.join("");
};
function reverse(arr, i, j) {
    while (i < j) {
        let temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
        i++;
        j--;
    }
}
